### PR TITLE
client/server tool invocations with `useChat` and `streamText` within Solidjs

### PR DESCRIPTION
**Pull request by Rajaniraiyn**
## Summary

Implements the `experimental_onToolCall` and `experimental_addToolResult` functionalities to the `ai/solid` module, based on the implementation in PR #1514. These enhancements streamline the process of handling client/server tool invocations in SolidJS applications.

## Example

- `examples/solidstart-openai/app/use-chat-client-tool`: confirmation dialog and tool that requires confirmation
- `examples/solidstart-openai/app/use-chat-tool-result`: display in-progress call & result for server side tool

### CODE REVIEW

**Review by lgrammel — APPROVED** [review]

**Comment by lgrammel** [comment]
Thank you @Rajaniraiyn this looks great!

File: packages/core/solid/use-chat.ts [hunk 1]
```diff
@@ -84,7 +85,18 @@ export function useChat({
   body,
   streamMode,
   generateId = generateIdFunc,
-}: UseChatOptions = {}): UseChatHelpers {
+}: Omit<UseChatOptions, 'api'> & {
+  api?: string;
```
Comment (lgrammel) [inline packages/core/solid/use-chat.ts:89]: This is React specific and can be reverted (no omit `api` and resetting it)

**Comment by lgrammel** [comment]
I'm running into a compiler issue with the index.tsx in the example: 
CleanShot 2024-05-28 at 10 41 23
Does this happen for you as well?

File: packages/core/solid/use-chat.ts (hunk 1 above)
Comment (Rajaniraiyn) [reply packages/core/solid/use-chat.ts]: reverted in https://github.com/vercel/ai/pull/1669/commits/e150d8a3d3c70ad99d070cd7aed76b765885f1c4

**Comment by Rajaniraiyn** [comment]
> I'm running into a compiler issue with the index.tsx in the example: CleanShot 2024-05-28 at 10 41 23 Does this happen for you as well?

Yes I got that issue as well. its happen to be a `ai` package installation issue. resolved in https://github.com/vercel/ai/pull/1669/commits/d1106589e2827c2c8b1f184c79cc93d4b6f10acd

**Comment by Rajaniraiyn** [comment]
@lgrammel, I noticed that the `handleInputChange` helper was missing for SolidJS. I have now implemented this functionality. You can find the changes in this commit: https://github.com/vercel/ai/pull/1669/commits/3391262df17ce890aa5aeae0f54a676e7ce88b49

File: packages/core/solid/use-chat.ts [hunk 2]
```diff
@@ -76,6 +81,7 @@ export function useChat({
   initialInput = '',
   sendExtraMessageFields,
   experimental_onFunctionCall,
+  experimental_onToolCall,
```
Comment (lgrammel) [inline packages/core/solid/use-chat.ts:84]: soon to be deprecated

File: packages/core/solid/use-chat.ts [hunk 3]
```diff
@@ -112,7 +126,14 @@ export function useChat({
...
   async function triggerRequest(
     messagesSnapshot: Message[],
-    { options, data }: ChatRequestOptions = {},
+    {
+      options,
+      data,
+      functions,
+      function_call,
+      tools,
+      tool_choice,
```
Comment (lgrammel) [inline packages/core/solid/use-chat.ts:135]: Soon to be depreated

**Review by lgrammel — COMMENTED** [review]
Can you split this up into 2 individual PRs?

- one for the experimental_addToolResult + example
- one for handleInputChange

I'd prefer not to bring the old function_call/tool_call things in here, they will soon be deprecated.

**Comment by Rajaniraiyn** [comment]
Sure. I'll make two separate PRs.

Should soon to be deprecated APIs marked as `@deprecated` or they be removed completely?

**Comment by lgrammel** [comment]
> Sure. I'll make two separate PRs.
> 
> Should soon to be deprecated APIs marked as `@deprecated` or they be removed completely?

I'm working on a replacement as we speak. I have not marked them deprecated yet as there is no replacement.

**Comment by Rajaniraiyn** [comment]
Closing this PR as there has been major refactor done on original codebase.
will open a new PR

**Pull request closed without merging**